Rework aarch64 stack frame implementation to use positive offsets.

cfallin opened the pull request:
This PR changes the aarch64 ABI implementation to use positive offsets
from SP, rather than negative offsets from FP, to refer to spill slots
and stack-local storage. This allows for better addressing-mode options,
and hence slightly better code: e.g., the unsigned scaled 12-bit offset
mode can be used to reach anywhere in a 32KB frame without extra
address-construction instructions, whereas negative offsets are limited
to a signed 9-bit unscaled mode (-256 bytes).

To enable this, the PR introduces a notion of "nominal SP offsets" as a
virtual addressing mode, lowered during the emission pass. The offsets
are relative to "SP after adjusting downward to allocate stack/spill
slots", but before pushing clobbers. This allows the addressing-mode
expressions to be generated before register allocation (or during it,
for spill/reload sequences).

To convert these offsets into *true* offsets from SP, we need to track
how much further SP is moved downward, and compensate for this. We do so
with "virtual SP offset adjustment" pseudo-instructions: these are seen
by the emission pass, and result in no instruction (0 byte output), but
update state that is now threaded through each instruction emission in
turn. In this way, we can push e.g. stack args for a call and adjust
the virtual SP offset, allowing reloads from nominal-SP-relative
spillslots while we do the argument setup with "real SP offsets" at the
same time.

As part of this PR, I also adjust how temporaries are used: I realized
(thanks to a comment in @alexcrichton's stack-limit PR) that x16 and x17
are the "interprocedural veneer temporaries", free to use within a
function body. These are now are spilltmp and "tmp2".

<!--

Please ensure that the following steps are all taken care of before submitting
the PR.

- [ ] This has been discussed in issue #..., or if not, please tell us why
  here.
- [ ] A short description of what this does, why it is needed; if the
  description becomes long, the matter should probably be discussed in an issue
  first.
- [ ] This PR contains test cases, if meaningful.
- [ ] A reviewer from the core maintainer team has been assigned for this PR.
  If you don't know who could review this, please indicate so. The list of
  suggested reviewers on the right can help you.

Please ensure all communication adheres to the [code of
conduct](https://github.com/bytecodealliance/wasmtime/blob/master/CODE_OF_CONDUCT.md).
-->

bjorn3 commented:
Will this still work if `alloca` is introduced in the future?

bnjbvr commented on cranelift/codegen/src/isa/aarch64/inst/mod.rs:
```diff
@@ -2616,42 +2625,58 @@ impl ShowWithRRU for Inst {
-                        ALUOp::Add64
+            &Inst::LoadAddr { rd, ref mem } => {
+                // TODO: we really should find a better way to avoid duplication of
+                // this logic bewteen `emit()` and `show_rru()` -- a separate 1-to-N
```
nit: between

bnjbvr commented on cranelift/codegen/src/isa/aarch64/abi.rs:
```diff
@@ -916,9 +871,15 @@ impl ABIBody for AArch64ABIBody {
         }
 
+        // N.B.: "nominal SP", which we use to refer to stackslots
+        // and spillslots, is *here*> We'll emit an adjustment meta-insn
```
nit: something is off with this comment?

bnjbvr commented on cranelift/codegen/src/isa/aarch64/inst/regs.rs:
```diff
@@ -125,21 +123,38 @@ pub fn writable_fp_reg() -> Writable<Reg> {
+/// Get a reference to the second temp register. We need this in some edge cases
+/// where we need both the spilltmp and another temporary.
+///
+/// We use x16 (aka IP0), the other "interprocedural"/linker-veneer scratch reg that is
```
nit: this is using x17 here

cfallin commented:
Thanks! Fixed the comment nits.

cfallin commented:
@bjorn3:
> Will this still work if alloca is introduced in the future?

That's a good point; we'll need to do something else in this case. I'll have to look at what other compilers do on aarch64 -- perhaps they just accept the cost of negative offsets from `fp`, or perhaps they put the fixed `fp`/`lr` frame below the stackslots/clobbers since aarch64 allows that flexibility... for now, though, since we don't have `alloca`, I think I'll want to save this as a future improvement.

bjorn3 commented:
> for now, though, since we don't have alloca, I think I'll want to save this as a future improvement.

:+1: I just wanted to prevent any decisions that would make it much harder to support `alloca`.

cfallin commented:
@bjorn3: yup, thanks for bringing it up, as it hadn't occurred to me! One solution (to record my thought for the future) might be to simply allocate a vreg as our "stackslot area base" and copy SP before we do any dynamic adjustments, and perhaps revert to this only if, on a scan, we see that an alloca is present.

cfallin commented:
Pushed a rebase over the recent file-split commit just now.

bnjbvr commented on cranelift/codegen/src/isa/aarch64/inst/args.rs:
```diff
@@ -137,11 +137,20 @@ pub enum MemArg {
     FPOffset(i64),
+
+    /// Offset from the "nominal stack pointer": the real SP, except when
+    /// it is temporarily adjusted downward to prepare for a call.
```
Since this comment will be largely read by everyone who runs into this kind of `MemArg`, it should be more precise. Can you explain what's above and below (closer to FP and further away from FP, to avoid the possible double interpretation of above/below)? Maybe draw some ASCII art to show a stack example and draw where the nominal SP is?

cfallin replied:
Done! Added a diagram in the main module comment for `abi.rs`; hopefully things are clearer now.

bnjbvr commented on cranelift/codegen/src/isa/aarch64/inst/args.rs:
```diff
@@ -112,7 +112,7 @@ pub enum MemLabel {
 #[derive(Clone, Debug)]
 pub enum MemArg {
-    Label(MemLabel),
+    // ---- Real ARM64 addressing modes ----
```
nit: remove `----` before and after text

bnjbvr commented on cranelift/codegen/src/isa/aarch64/inst/args.rs:
```diff
@@ -137,11 +137,20 @@ pub enum MemArg {
     UnsignedOffset(Reg, UImm12Scaled),
 
-    /// Offset from the stack pointer. Lowered into a real amode at emission.
+    /// ---- virtual addressing modes that are lowered at emission time ----
```
ditto

bnjbvr commented on cranelift/codegen/src/isa/aarch64/abi.rs:
```diff
@@ -800,47 +751,51 @@ impl ABIBody for AArch64ABIBody {
-        let fp_off: i64 = -(self.stackslots_size as i64) - (8 * islot) - ty_size as i64;
-        load_stack_from_fp(fp_off, into_reg, ty)
+        let spill_off = islot * 8;
+        // Offset from nominal-SP.
```
nit: maybe remove this comment? the `MemArg` enum variant below + above comment make it pretty clear it's an offset from nominal-sp.

bnjbvr commented on cranelift/codegen/src/isa/aarch64/abi.rs:
```diff
@@ -800,47 +751,51 @@ impl ABIBody for AArch64ABIBody {
         }
     }
 
     // Load from a spillslot.
```
preexisting: while you're around, can you make this a doc comment please?

bnjbvr commented on cranelift/codegen/src/isa/aarch64/abi.rs:
```diff
@@ -916,9 +871,16 @@ impl ABIBody for AArch64ABIBody {
+        // and spillslots, is *here* (the value of SP at this program point).
+        // If we push any clobbers below, we emit a virtual-SP adjustment
+        // meta-instruction so that the nominal-SP references behave as if SP
+        // were still at this point.
```
As said below, I think this comment should be placed closer to `MemArg::NominalSPOffset`, with a drawing of what it should look like.

bnjbvr commented on cranelift/codegen/src/isa/aarch64/abi.rs:
```diff
@@ -1139,14 +1115,20 @@ impl AArch64ABICall {
 
 fn adjust_stack(amt: u64, is_sub: bool) -> Vec<Inst> {
     if amt > 0 {
+        let adj = if is_sub { amt as i64 } else { -(amt as i64) };
```
nit: can you name this `sp_adjustment` and also rename `amt` into `amount`?

bnjbvr commented on cranelift/codegen/src/machinst/abi.rs:
```diff
@@ -97,8 +97,8 @@ pub trait ABIBody {
-    /// Returns the full frame size for the given function, after prologue emission has run. This
-    /// comprises the spill space, incoming argument space, alignment padding, etc.
+    /// Returns the size of a function call frame (including return address and FP) for this
+    /// function's body.
```
This comment is less precise than the previous one, can you tweak it to tell what this number includes?

bnjbvr commented on cranelift/codegen/src/machinst/mod.rs:
```diff
@@ -215,7 +215,9 @@ pub enum MachTerminator<'a> {
-    fn emit(&self, code: &mut O, flags: &Flags);
+    fn emit(&self, code: &mut O, flags: &Flags, state: &mut Self::State);
+    /// Persistent state carried across `emit` invocations.
+    type State: Default + Clone + Debug;
```
nit: can you put this before the `emit` function which uses it, to make reading of this code more linear?

bjorn3 commented on cranelift/codegen/src/isa/aarch64/abi.rs:
```diff
@@ -800,47 +751,51 @@ impl ABIBody for AArch64ABIBody {
-        let fp_off: i64 = -(self.stackslots_size as i64) + stack_off + (offset as i64);
-        load_stack_from_fp(fp_off, into_reg, ty)
+        let sp_off: i64 = stack_off + (offset as i64);
+        debug!("load_stackslot: slot {} -> sp_off {}", slot, sp_off);
```
It can become a `trace!`.

cfallin commented:
Thanks! Updated, PTAL.

> I think `lower_addr` may reuse the temporary spill registers as well, can you please take a look?

Fixed (I assume you meant `load_addr`?) -- using `tmp2` now and added a comment to justify why it's safe.

cfallin commented:
And, I should note I built a SpiderMonkey JS shell with this commit vendored in, and all jit-tests pass; so I think the Baldrdash-related stuff is still correct.

bnjbvr commented on cranelift/codegen/src/isa/aarch64/inst/args.rs:
```diff
@@ -152,6 +174,14 @@ impl MemArg {
     }
 
+    /// Return a reg if this memory address is just a register with no other offset.
+    pub fn is_reg_no_offset(&self) -> Option<Reg> {
```
This is actually unused in this PR, so it could be removed? (unless you plan to use it soon later)

(`pub(crate)` would show you a warning if it's effectively unused)

cfallin replied:
Removed, thanks!
